src/app/modules/services/pull-action-protocol/pull-action-protocol.service.ts

```diff
@@ -0,0 +1,230 @@
+  backedUpResources: Array<BackupEntry>;
+}
+
+export class BackupEntrySet {
+  entries: BackupEntry[] = [];
+
+  add(additionalEntries: BackupEntry[]): void {
+    const newEntries = additionalEntries.filter(additionalEntry => !this.entries.find(entry => this.equals(entry, additionalEntry)));
+    this.entries = this.entries.concat(newEntries);
+  }
+
+  equals(entry1: BackupEntry, entry2: BackupEntry): boolean {
+    return entry1.backupResource === entry2.backupResource
+      && entry1.resource === entry2.resource;
+  }
+
+  toArray(): BackupEntry[] {
+    return this.entries;
+  }
+
+}
+
+/** Wrap a simple action (e.g. copy, list-files etc.) into a protocol that does the following:
+
+    1. pull, collect interesting changes in diffs, that may be of interest to the ui
+       - if the pull fails, report a pull failure and exit
+    2. execute the repository action
+       - if the action could not be executed by the backend, and the backend indicates to repull,
+         undo the action (user of the protocol should start over at step 1, collecting additional! diffs of interest)
+       - if an error occured during the execution of the action,
+         undo the action (user of the protocol should start over at step 1, collecting additional! diffs of interest)
+       - if errors repeatedly occur, even though additional pulls did not change the repository (since it was up to date),
+         fail the whole action
+
+
+    usage:
+      const pullActionProtocol = new PullActionProtocol(...);
+      while (pullActionProtocol.retryExecution()) {
+        await pullActionProtocol.execute();
+        // do some logging?
+      }
+      // inform about changes of interest
+      // pullActionProtocol.changedResourcesSet: contains all files that changed due to executed pulls.
+      //                    changedResourcesSet is filled with filesInterestedInContentChanges (see ctor)
+      // pullActionProtocol.backedUpResourcesSet: contains files (and the corresponding backup files) that changed due to executed pulls.
+      //                    backedUpResourcesSet is filled with fileListToBackupOnChange (see ctor)
+      // if pull result in changes to files in the criticalFilesOfInterest list, the process stops with an error (pullActionProcotol.result)
+
+      if (pullActionProtocol.result instanceof Error) {
+        throw pullActionProcotol.result;
+      } else if (pullActionProtocol.result instanceof Conflict) {
+        ...
+      } else {
+        return pullActionProcotol.result;
+      }
+  */
+export class PullActionProtocol<T> {
+  changedResourcesSet = new Set<string>();
+  backedUpResourcesSet = new BackupEntrySet();
+  executedRetries: number;
+  consecutiveExecutedRetriesWithoutDiff: number;
+  httpClient: HttpClient;
+  result: T | Conflict | Error;
+
+  readonly filesOfInterest: string[];
+  readonly dirtyFilesOfInterest: string[];
+
+  constructor(private httpProvider: HttpProviderService, private serviceUrl: string,
```

Comment:
While reading the tests I thought it might be nicer if the constructor accepted a single object (because then you can call it with an object literal that includes the names of the properties being passed in -- otherwise its hard to see which parameter means what…). The downside is that you'd have to declare the fields separately, and assign them in the constructor's body. Not sure which variant is better :\